core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java

```diff
@@ -215,6 +223,83 @@
+                                                                  Class<E> cast) {
+        return (E) getAnnotationField(annotation, fieldName);
+    }
+
+    private static <T extends Annotation> Object getAnnotationField(T annotation, String fieldName) {
+        try {
+            return annotation.getClass().getMethod(fieldName).invoke(annotation);
+        } catch (Exception ex) {
+            throw new IllegalStateException(
+                    String.format("%s() method invocation throws an exception", fieldName), ex);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T extends Annotation> List<T> findRepeatableAnnotations(AnnotatedElement element,
+                                                                            Class<T> annotationType) {
+        requireNonNull(element, "element");
+        requireNonNull(annotationType, "annotationType");
+        final Class<? extends Annotation> repeatable = getRepeatable(annotationType).orElseThrow(
+                () -> new IllegalArgumentException("annotationType is not a repeatable annotation."));
+        return getAllAnnotations(element).stream()
+                                         .filter(a -> a.annotationType() == annotationType ||
+                                                      a.annotationType() == repeatable)
+                                         .flatMap(annotation -> {
+                                             if (annotation.annotationType() == repeatable) {
+                                                 return Arrays.stream(
+                                                         (T[]) getAnnotationField(annotation, "value"));
+                                             }
+                                             return Stream.of((T) annotation);
+                                         }).collect(toImmutableList());
+    }
+
+    private static <T extends Annotation> void setAdditionalHeader(HttpHeaders headers,
+                                                                   AnnotatedElement element,
+                                                                   String clsAlias,
+                                                                   String level,
+                                                                   Class<T> annotation) {
+        requireNonNull(headers, "headers");
+        requireNonNull(element, "element");
+        requireNonNull(level, "level");
+
+        final Set<String> addedHeaderSets = new HashSet<>();
+        findRepeatableAnnotations(element, annotation).forEach(header -> {
+            final String name = getAnnotationField(header, "name", String.class);
+            final String[] value = getAnnotationField(header, "value", String[].class);
+
+            if (addedHeaderSets.contains(name)) {
+                logger.warn("The additional {} named '{}' is already added at the same {} level." +
```

Comment:
- `is already added at the same {} level. It will be ignored.` -> `is set at the same {} level already; ignoring.`
- How about telling the class name (and method) which has the problem as well?  e.g. in the format of `<className>.<methodName>()`